wlog/auditlog/audit3log/logger.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2018 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package audit3log
+
+import (
+	"io"
+
+	"github.com/palantir/witchcraft-go-logging/wlog"
+)
+
+type AuditResultType string
+type AuditProducerType string
+type AuditSensitivityType string
+
+type AuditOrganizationType struct {
+	ID     string `json:"id"`
+	Reason string `json:"reason"`
+}
+
+type AuditContextualizedUserType struct {
+	UID       string   `json:"uid"`
+	UserName  string   `json:"userName"`
+	FirstName string   `json:"firstName"`
+	LastName  string   `json:"lastName"`
+	Groups    []string `json:"groups"`
+	Realm     string   `json:"realm"`
+}
+
+type AuditSensitivityTaggedValueType struct {
+	Level   []AuditSensitivityType `json:"level"`
+	Payload interface{}            `json:"payload"`
+}
+
+const (
+	AuditResultSuccess        AuditResultType      = "SUCCESS"
+	AuditResultUnauthorized   AuditResultType      = "UNAUTHORIZED"
+	AuditResultError          AuditResultType      = "ERROR"
+	AuditProducerServer       AuditProducerType    = "SERVER"
+	AuditProducerClient       AuditProducerType    = "CLIENT"
+	AuditSensitivityMetadata  AuditSensitivityType = "Metadata"
+	AuditSensitivityUserInput AuditSensitivityType = "UserInput"
+	AuditSensitivityData      AuditSensitivityType = "Data"
+)
+
+type Logger interface {
+	Audit(name string, result AuditResultType, deployment string, host string, product string, productVersion string, params ...Param)
```

Comment:
We discussed making these immutable fields (deployment, host, product, version) part of the logger's constructor and stored as struct fields rather than requiring applications to plumb these to every call site. Alternatively they could all be params added to the context, but enforcing that they are required is harder.
It also looks like service, stack, and environment are currently params, which is inconsistent.

What do you think?

Reply:
I will get started on making the immutable fields part of the logger constructor today!

re: "It also looks like service, stack, and environment are currently params, which is inconsistent." - I included these as params as they are included as optional within the SLS spec. I agree that this is inconsistent though and would be an easy add to the constructor struct. Do you have thoughts on whether these should be included as required or optional within this struct?